tt_metal/llrt/tt_cluster.cpp

```diff
@@ -582,82 +585,92 @@ std::unordered_map<chip_id_t, std::vector<CoreCoord>> Cluster::get_ethernet_core
     return connected_chips;
 }
 #define MAX_TUNNEL_DEPTH 4
-std::vector<std::vector<chip_id_t>> Cluster::get_tunnels_from_mmio_device(chip_id_t mmio_chip_id) const {
-    std::vector<std::vector<chip_id_t>> tunnels_from_mmio = {};
-    const auto &all_eth_connections = this->cluster_desc_->get_ethernet_connections();
-    TT_ASSERT(this->cluster_desc_->is_chip_mmio_capable(mmio_chip_id));
-
-    if (all_eth_connections.find(mmio_chip_id) == all_eth_connections.end()) {
-        return {};
-    }
+void Cluster::set_tunnels_from_mmio_device() {
+    for (const auto &[mmio_chip_id, physical_chip_id] : this->cluster_desc_->get_chips_with_mmio()) {
+        std::vector<std::vector<chip_id_t>> tunnels_from_mmio = {};
+        const auto &all_eth_connections = this->cluster_desc_->get_ethernet_connections();
+        TT_ASSERT(this->cluster_desc_->is_chip_mmio_capable(mmio_chip_id));
+
+        if (all_eth_connections.find(mmio_chip_id) == all_eth_connections.end()) {
+            this->tunnels_from_mmio_device.insert({mmio_chip_id, {}});
+        }
 
-    std::set<chip_id_t> device_ids = get_devices_controlled_by_mmio_device(mmio_chip_id);
-    device_ids.erase(mmio_chip_id);
+        std::set<chip_id_t> device_ids = get_devices_controlled_by_mmio_device(mmio_chip_id);
+        device_ids.erase(mmio_chip_id);
```

Comment:
can we clean up this method or add some helper methods..? we're erasing device ids in a few places and seems kind of error-prone.

Reply:
My change here is a clang-format. I think this should maybe be using an object to track visited chips.

Reply:
okay, then it's out of scope for this PR. We should come back to this and clean it up at some point but it's fine for now.